File: app/src/main/java/com/mxt/anitrend/model/api/interceptor/NetworkCacheInterceptor.java
package com.mxt.anitrend.model.api.interceptor;

import android.content.Context;
import android.support.annotation.NonNull;
import com.mxt.anitrend.util.CompatUtil;

import java.io.IOException;
import java.util.concurrent.TimeUnit;

import okhttp3.CacheControl;
import okhttp3.Interceptor;
import okhttp3.Response;

/**
 * Created by max on 2017/07/17.
 * Network cache injector interceptor
 */

public class NetworkCacheInterceptor implements Interceptor {

    private static final String CACHE_CONTROL = "Cache-Control";
    private Context mContext;
    private boolean forceCache;

    // re-write response header to force use of cache
    private CacheControl cacheControl = new CacheControl.Builder()
            .maxAge(6, TimeUnit.HOURS)
            .build();

    public NetworkCacheInterceptor(Context mContext) {
        this.mContext = mContext;
    }

    public NetworkCacheInterceptor(Context mContext, boolean forceCache) {
        this.mContext = mContext;
        this.forceCache = forceCache;
    }

    @Override
    public Response intercept(@NonNull Chain chain) throws IOException {
        Response response = chain.proceed(chain.request());
        if(!CompatUtil.isOnline(mContext) || !forceCache)
            return response.newBuilder().header(CACHE_CONTROL,
                    cacheControl.toString()).build();
        return response;
    }
}

File: app/src/main/java/com/mxt/anitrend/model/api/interceptor/NetworkCacheInterceptor.kt
package com.mxt.anitrend.model.api.interceptor

import android.content.Context
import com.mxt.anitrend.util.CompatUtil
import okhttp3.CacheControl
import okhttp3.Interceptor
import okhttp3.Response
import java.io.IOException
import java.util.concurrent.TimeUnit

/**
 * Created by max on 2017/07/17.
 * Network cache injector interceptor
 */

class NetworkCacheInterceptor(private val context: Context?, private val forceCache: Boolean = false) : Interceptor {

    // re-write response header to force use of cache
    private val cacheControl by lazy {
        CacheControl.Builder()
                .maxAge(4, TimeUnit.HOURS)
                .build()
    }

    @Throws(IOException::class)
    override fun intercept(chain: Interceptor.Chain): Response {
        val response = chain.proceed(chain.request())
        return if (!CompatUtil.isOnline(context) || !forceCache) response.newBuilder().header(CACHE_CONTROL,
                cacheControl.toString()).build() else response
    }

    companion object {
        private const val CACHE_CONTROL = "Cache-Control"
    }
}
